### PR TITLE
[2] Supporting custom key encoders for key-value stages in mantis

### PR DESCRIPTION
**Pull request by hmitnflx**
- allow non-string codecs for input in mantis group to scalar stages
- pass the custom keyencoder to be used during deserialization

### Context
Right now the keys inside MantisGroup are assumed to be String always.
This might not be the case in the future and so let's refactor to support
other encoders. Also see companion PRs #212 and #214 

### Checklist

- [x] `./gradlew build` compiles code correctly
- [x] Added new tests where applicable
- [x] `./gradlew test` passes all tests
- [x] Extended README or added javadocs where applicable
- [x] Added copyright headers for new files from `CONTRIBUTING.md`

### CODE REVIEW

File: mantis-runtime/src/main/java/io/mantisrx/runtime/GroupToGroup.java [hunk 1]
```diff
@@ -49,6 +50,11 @@
 
     GroupToGroup(GroupComputation<K1, T, K2, R> computation,
                  Config<K1, T, K2, R> config, Codec<T> inputCodec) {
+        this(computation, config, (Codec<K1>) Codecs.string(), inputCodec);
+    }
+
+    GroupToGroup(GroupComputation<K1, T, K2, R> computation,
```
Comment (hmitnflx) [inline mantis-runtime/src/main/java/io/mantisrx/runtime/GroupToGroup.java:56]: This actually results in a diamond inheritance. I'm thinking that avoiding `KeyValueInputStageConfig` and keeping the key encoder in `Stages` class is along with StageConfig is a good compromise. 
I'm unsure if I should make that optional or nullable, though. Thoughts?

File: mantis-runtime/src/main/java/io/mantisrx/runtime/ScalarStages.java [hunk 2]
```diff
@@ -27,17 +27,17 @@
 public class ScalarStages<T> extends Stages<T> {
 
     ScalarStages(SourceHolder<?> source, StageConfig<?, ?> stage, Codec<T> inputCodec) {
-        super(source, stage, inputCodec);
+        super(source, stage, null, inputCodec);
     }
 
     ScalarStages(Stages<?> self, StageConfig<?, ?> stage, Codec<T> inputCodec) {
-        super(self.getSource(), self.getStages(), stage, inputCodec);
+        super(self.getSource(), self.getStages(), stage, null, inputCodec);
```
Comment (hmitnflx) [inline mantis-runtime/src/main/java/io/mantisrx/runtime/ScalarStages.java:34]: thoughts on Optional vs nullable here?

**Comment by calvin681** [comment]
Can you put in the PR description the motivation for non-string keys?

File: mantis-runtime/src/main/java/io/mantisrx/runtime/KeyValueInputStageConfig.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+package io.mantisrx.runtime;
+
+import io.mantisrx.common.codec.Codec;
+import io.mantisrx.common.codec.Codecs;
+import io.mantisrx.runtime.parameter.ParameterDefinition;
+import java.util.List;
+
+public class KeyValueInputStageConfig<K, T, R> extends StageConfig<T, R> {
```
Comment (calvin681) [inline mantis-runtime/src/main/java/io/mantisrx/runtime/KeyValueInputStageConfig.java:24]: why not add InputKeyCodec and OutputKeyCodec to StageConfig directly? is this class necessary?

Comment (hmitnflx) [reply mantis-runtime/src/main/java/io/mantisrx/runtime/KeyValueInputStageConfig.java:24]: I can definitely do that. That's exactly my question [here](https://github.com/Netflix/mantis/pull/218#discussion_r863155877)

**Review by calvin681 — APPROVED** [review]

**Pull request merged**